src/main/kotlin/org/elm/lang/core/types/Ty.kt

```diff
@@ -59,11 +59,14 @@ data class TyTuple(val types: List<Ty>, override val alias: AliasInfo? = null) :
  * @property baseTy The type of the base record identifier, if there is one. Non-null for field
  *   accessors, record with base identifiers etc. that match a subset of record fields
  * @property alias The alias for this record, if there is one. Used for rendering and tracking record constructors
+ * @property fieldReferences A map of field name to the psi element that defines them; used for
+ *   reference resolve, does not affect equality
  */
 data class TyRecord(
         val fields: Map<String, Ty>,
         val baseTy: Ty? = null,
-        override val alias: AliasInfo? = null
+        override val alias: AliasInfo? = null,
+        val fieldReferences: RecordFieldReferenceTable = RecordFieldReferenceTable()
```

Comment:
I really don't like this hack of having a mutable collection of PsiElements tacked on to a Ty, but it is the best solution I could come up with since it works well with all the unification code. I tried to mitigate the potential downsides by freezing this table when a ty is cached to prevent modification of cached values.